checker/src/org/checkerframework/checker/lock/LockVisitor.java

```diff
@@ -354,9 +606,342 @@ protected boolean skipContractCheck(Tree tree, FlowExpressions.Receiver expr, Fl
+                    TypeElement classElem = TreeUtils
+                            .elementFromDeclaration(enclosingClass);
+                    Node receiver = new ImplicitThisLiteralNode(classElem.asType());
+
+                    Receiver internalReceiver = FlowExpressions.internalReprOf(atypeFactory,
+                            receiver);
+
+                    flowExprContext = new FlowExpressionContext(
+                            internalReceiver, null, checker.getContext());
+                }
+                else if (nodeNode instanceof ArrayAccessNode) {
+                    // Adapted from FlowExpressionParseUtil.buildFlowExprContextForUse
+
+                    Receiver internalReceiver = FlowExpressions.internalReprOfArrayAccess(atypeFactory,
+                        (ArrayAccessNode) nodeNode);
+
+                    flowExprContext = new FlowExpressionContext(
+                            internalReceiver, null, checker.getContext());
+                } else if (nodeNode instanceof ExplicitThisLiteralNode ||
+                           nodeNode instanceof ImplicitThisLiteralNode ||
+                           nodeNode instanceof ThisLiteralNode) {
+                    Receiver internalReceiver = FlowExpressions.internalReprOf(atypeFactory, nodeNode, false);
+
+                    flowExprContext = new FlowExpressionContext(
+                            internalReceiver, null, checker.getContext());
+                }
+            }
+
+            if (flowExprContext != null) {
+                FlowExpressions.Receiver expr = null;
+                try {
+                    CFAbstractStore<?, ?> store = atypeFactory.getStoreBefore(tree);
+
+                    String s = expression.trim();
+                    Pattern selfPattern = Pattern.compile("^(this)$");
+                    Matcher selfMatcher = selfPattern.matcher(s);
+                    if (selfMatcher.matches()) {
+                        s = flowExprContext.receiver.toString(); // it is possible that s == "this" after this call
+                    }
+
+                    expr = FlowExpressionParseUtil.parse(expression,
+                            flowExprContext, getCurrentPath());
+
+                    if (expr == null) {
+                        // TODO: Wrap the following 'itself' handling logic into a method that calls FlowExpressionParseUtil.parse,
+                        // and call that method instead of FlowExpressionParseUtil.parse immediately above.
+
+                        /** Matches 'itself' - it refers to the variable that is annotated, which is different from 'this' */
+                        Pattern itselfPattern = Pattern.compile("^itself$");
+                        Matcher itselfMatcher = itselfPattern.matcher(expression.trim());
+
+                        if (itselfMatcher.matches()) { // There is no variable, class, etc. named "itself"
+                            if (itselfIsTheReceiverNode && methodCall) {
```

Comment:
I don't understand why `itselfIsTheReceiverNode` is needed.  If `tree` is a method call, doesn't "itself" always refer to the receiver of the method call?